src/background/modules/AutocorrectHandler.js

```diff
@@ -26,6 +26,76 @@ let antipatterns = [];
 // Adapted from: https://github.com/mozilla/webextension-polyfill/blob/master/src/browser-polyfill.js
 const IS_CHROME = Object.getPrototypeOf(browser) !== Object.prototype;
 
+/**
+ * Traverse Trie tree of objects to create RegEx.
+ *
+ * @param {Object.<string, Object|boolean>} obj
+ * @returns {string}
+ */
+function pattern(obj) {
+    const alt = [];
+    const cc = [];
+
+    // Escape special characters
+    const regExSpecialChars = /[.*+?^${}()|[\]\\]/gu;
+
+    for (const char in obj) {
+        if (char) {
+            const achar = char.replace(regExSpecialChars, "\\$&");
+
+            if (!("" in obj[char] && Object.keys(obj[char]).length === 1)) {
+                const recurse = pattern(obj[char]);
+                alt.push(recurse + achar);
+            } else {
+                cc.push(achar);
+            }
+        }
+    }
+
+    if (cc.length) {
+        alt.push(cc.length === 1 ? cc[0] : `[${cc.join("")}]`);
+    }
+
+    let result = alt.length === 1 ? alt[0] : `(?:${alt.join("|")})`;
+
+    if ("" in obj) {
+        if (cc.length || alt.length > 1) {
+            result += "?";
+        } else {
+            result = `(?:${result})?`;
+        }
+    }
+
+    return result;
+}
+
+/**
+ * Convert autocorrections into Trie tree of objects.
+ *
+ * @param {string[]} arr
+ * @returns {string}
+ */
+function createRegEx(arr) {
+    const tree = {};
+
+    for (const s of arr.sort((a, b) => b.length - a.length)) {
```

Comment:
This creates the three? Then maybe extract it into `createTree`. As mentioned above, maybe a JS-class trieThree wold be useful. 

I found this example, which looks quite clean and could be used (or adapted) if the license is clear: https://gist.github.com/tpae/72e1c54471e88b689f85ad2b3940a8f0 (see also the comments which look more optimized)

(maybe we could remove the parts we do not need of course)

---

More inspiration:
* TS implementation: https://github.com/leekevinyg/js-trie
* learning article: https://dev.to/namdx1987/js-simple-trie-implementation-n23

Reply:
I could rename this function to `createTree` as suggested and then rename the above `pattern` function to `createRegEx`, if you think that would be more clear.

Those examples just create a Trie tree and not the needed RegEx. They also do not handle the intricacies of our use case. Specifically, we need to create the Trie tree in reverse, as the resulting RegEx is right anchored since we are matching the text before the carrot as the user is typing.

For reference, my `pattern` function was loosely adapted from this Python code: https://gist.github.com/EricDuminil/8faabc2f3de82b24e5a371b6dc0fd1e0, which I ported to JS and heavily refactored/optimized.